PWGHF: first checked implementation of Xic->pKpi filtering

mfaggin opened the pull request.

Commit:
PWGHF: first checked code of Xic->pKpi filtering

mfaggin commented:
Linked PR on validation framework: https://github.com/AliceO2Group/Run3Analysisvalidation/pull/158

mfaggin commented:
Linked PR in validation framework: https://github.com/AliceO2Group/Run3Analysisvalidation/pull/158

mfaggin committed:
Merge branch 'dev' into mfaggin_XicTopKpi

mfaggin commented:
@ginnocen @vkucera @DelloStritto few required fixes implemented

mfaggin commented:
@ginnocen switched to "ready for review" status

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -382,6 +386,38 @@ auto InvMassLcpiKp(const T& candidate)
+
+template <typename T>
+auto InvMassXicpiKp(const T& candidate)
+{
+  return InvMassLcpiKpi(candidate);
+}
```
@mfaggin The `InvMassLcpiKpi` function does not exist. Does O2 compile for you with this typo?

mfaggin replied:
@vkucera indeed it does not, sorry for the typo, I fixed it

mfaggin committed:
Update HFSecondaryVertex.h

DelloStritto commented on Analysis/Tasks/PWGHF/HFXicCandidateSelector.cxx:
```diff
@@ -0,0 +1,447 @@
+        }
+      }
+
+      if ((pidXicPKPi == 1) && topolXicpKpi) {
```
Hi @mfaggin  . You are not including the case in which pidXicPKPi or pidXicPiKP = -1 (i.e. no PID information).
At the moment you are rejecting them and this is a very tight selection.

mfaggin replied:
Hi @DelloStritto , I think this is not true, because pidXicPKPi and pidXicPiKP are set to 1 both in case the returned value from selectionPID is +1 or -1 (look at lines 422 for example: I am checking the absolute value). Do you agree or am I wrong?

DelloStritto replied:
Ah, I see. You are requiring the abs. But at this point shoudn't be the same for the kaon? (L. 421)

mfaggin replied:
Hi @DelloStritto , indeed you are right! Thanks a lot, I fixed it.

mfaggin committed:
Update HFXicCandidateSelector.cxx

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFCandidateSelectionTables.h:
```diff
@@ -39,4 +39,14 @@ DECLARE_SOA_COLUMN(IsSelJpsiToEE, isSelJpsiToEE, int);
+{
+namespace hf_selcandidate_xic
+{
+DECLARE_SOA_COLUMN(IsSelXicpKpi, isSelXicpKpi, int);
+DECLARE_SOA_COLUMN(IsSelXicpiKp, isSelXicpiKp, int);
+} // namespace hf_selcandidate_xic
+DECLARE_SOA_TABLE(HFSelXicpKpiCandidate, "AOD", "HFSELXICCAND", hf_selcandidate_xic::IsSelXicpKpi, hf_selcandidate_xic::IsSelXicpiKp);
```
These should be `IsSelXicToPKPi` etc. to follow the camelCase convention and to match the `DecayType` names.
@DelloStritto could you please add these in your Lc PR as well?

DelloStritto replied:
Yes, I will. Thanks

mfaggin replied:
I am going to do it as well, I will let you know when I conclude.

vkucera replied:
Thanks

mfaggin replied:
@vkucera I implemented the requested syntax fixes.

vkucera replied:
@mfaggin Thanks. Can you fix the rest as well?

mfaggin replied:
@vkucera done also for other Xic methods (see commit 38af94a , clang checks are queued, I will fix possible syntax errors tomorrow, sorry I need to leave now)

mfaggin replied:
@vkucera I see that the clang checks are ok now, so the requested modifications are in place.

Commit:
Fix code with camelCase syntax conventions

vkucera commented:
@mfaggin @DelloStritto @ginnocen If I understand correctly, the only difference between the Lc and Xic selector should eventually be the configuration (including the mass cut) and the output flag column.
Why don't we try to reuse as many functions as possible from the existing Lc selector?

mfaggin commented:
> @mfaggin @DelloStritto @ginnocen If I understand correctly, the only difference between the Lc and Xic selector should eventually be the configuration (including the mass cut) and the output flag column.
> Why don't we try to reuse as many functions as possible from the existing Lc selector?

IMHO, this should be avoided at this stage. My implementation of the Xic is indeed based on the Lc task, since we are talking about particles that are reconstructed in the same decay channel. However, being two different particles with different kinematics, I would not force them to be equal or strongly correlated for the moment, because for real analysis things may differentiate. I agre with you for methods like the one on the pKpi invariant mass calculations, but not for all methods.

vkucera commented:
@mfaggin Is there any topological cut that could ever be applied only for Lc and not for Xic?

mfaggin commented:
@vkucera at this stage I do not see indeed, but I do not have such a long-time view to confirm it at 100%. I can try to have a look if such an operation would be too constraining or not, but I do not have too much time to dedicate on this business in a short time-scale. Do you think it's crucial for this PR to be approved? I may think to address this topic after this PR merging, so that something is already in place for Xic.

Commit:
Fix code syntax conventions for Xic methods

mfaggin commented:
Hi @vkucera , I see conflicts in Analysis/Tasks/PWGHF/CMakeLists.txt that probably depend on the fact that in the meanwhile something changed in the dev branch (also for the validation framework). I think it's better if you fix it as you prefer, according to the new conventions, in both cases. Do you agree?

vkucera commented:
> Hi @vkucera , I see conflicts in Analysis/Tasks/PWGHF/CMakeLists.txt that probably depend on the fact that in the meanwhile something changed in the dev branch (also for the validation framework). I think it's better if you fix it as you prefer, according to the new conventions, in both cases. Do you agree?

Hi @mfaggin , those should be just conflicts due to the addition of the MC validation task. You should just rebase, resolve the conflicts and push again.

mfaggin commented:
> > Hi @vkucera , I see conflicts in Analysis/Tasks/PWGHF/CMakeLists.txt that probably depend on the fact that in the meanwhile something changed in the dev branch (also for the validation framework). I think it's better if you fix it as you prefer, according to the new conventions, in both cases. Do you agree?
> 
> Hi @mfaggin , those should be just conflicts due to the addition of the MC validation task. You should just rebase, resolve the conflicts and push again.

Hi @vkucera , I manually resolved the conflicts directly from the git interface, please check if my implementations are ok.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -198,6 +198,15 @@ struct HFCandidateCreator3ProngMC {
+      // Ξc± → p± K∓ π±
+      if (flag == 0) {
+        //Printf("Checking Ξc± → p± K∓ π±");
+        auto indexRecXic = RecoDecay::getMatchedMCRec(particlesMC, std::move(arrayDaughters), 4232, array{+kProton, -kKPlus, +kPiPlus}, true, &sign);
```
`std::move` must always go only once for the last usage of `arrayDaughters`, so it has to be removed from `auto indexRecLc = `.

mfaggin replied:
I removed it from the `RecoDecay::getMatchedMCRec` call for Lc

vkucera replied:
You don't seem to need to store the return value of `RecoDecay::getMatchedMCRec`, so you can just directly check it in the `if` condition, as done for the D0.
Same for `isMatchedMCGen`.

vkucera commented on Analysis/Tasks/PWGHF/HFCandidateCreator3Prong.cxx:
```diff
@@ -239,6 +248,15 @@ struct HFCandidateCreator3ProngMC {
+      // Ξc± → p± K∓ π±
+      if (flag == 0) {
+        //Printf("Checking Ξc± → p± K∓ π±");
+        auto indexRecXic = RecoDecay::isMatchedMCGen(particlesMC, particle, 4232, array{+kProton, -kKPlus, +kPiPlus}, true, &sign);
+        if (indexRecXic > -1) {
```
This is wrong. Check how it is done for Λc±.

mfaggin replied:
Sorry, my implementation was probably a refuse from the `RecoDecay::getMatchedMCRec` call. I modified the name of the returned value, as well as the condition inside the `if` statement.

vkucera commented on Analysis/Tasks/PWGHF/HFXicToPKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,447 @@
+
+/// \file HFXicToPKPiCandidateSelector.cxx
+/// \brief Xic->pKpi selection task.
+/// \note Inspired from HFXicToPKPiCandidateSelector.cxx
```
Self-inspired? :-D

mfaggin replied:
... of course not :) changed

vkucera commented on Analysis/Tasks/PWGHF/taskXic.cxx:
```diff
@@ -0,0 +1,171 @@
+struct TaskXicMC {
+  HistogramRegistry registry{
+    "registry",
+    {{"hPtRecSig", "3-prong candidates (rec. matched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtRecBg", "3-prong candidates (rec. unmatched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGen", "3-prong candidates (gen. matched);#it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hPtGenSig", "3-prong candidates (rec. matched);#it{p}_{T}^{gen} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hCPARecSig", "3-prong candidates (rec. matched);cosine of pointing angle;entries", {HistType::kTH1F, {{110, -1.1, 1.1}}}},
+     {"hCPARecBg", "3-prong candidates (rec. unmatched);cosine of pointing angle;entries", {HistType::kTH1F, {{110, -1.1, 1.1}}}},
+     {"hEtaRecSig", "3-prong candidates (rec. matched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hEtaRecBg", "3-prong candidates (rec. unmatched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}},
+     {"hEtaGen", "3-prong candidates (gen. matched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}}}};
```
Adjust the titles for MC particle histograms. (See `taskD0.cxx`.)

mfaggin replied:
I adjusted the histogram titles as required.
@vkucera

vkucera reviewed (changes requested):
As discussed, the selector should use functions from the Lc selector, but that can be done in the next steps.

Commit:
PWGHF: further fixes for Xic to pKpi filtering and analysis

 - (HFCandidateCreator3Prong.cxx) remove refuse std::move, forgot in the middle of the for loop
 - (HFCandidateCreator3Prong.cxx) fix if statement condition for isMatchedMCGen
 - (taskXic.cxx) fix MC histogram titles

Commit:
PWGHF: avoid storing return values of RecoDecay::getMatchedMCRec and isMatchedMCGen for Xic->pKpi

mfaggin commented:
@vkucera I followed your instructions.

vkucera commented:
The code looks fine now.
Just the variable names need to be fixed to comply with the conventions.

mfaggin commented:
> The code looks fine now.
> Just the variable names need to be fixed to comply with the conventions.

Hi @vkucera , which variables are you referring to? If I am not wrong, the variable names in the `HFXicToPKPiCandidateSelector.cxx` are equal to the ones in `HFLcToPKPiCandidateSelector.cxx`, but I am not sure if you are referring to them.

vkucera commented:
> > The code looks fine now.
> > Just the variable names need to be fixed to comply with the conventions.
> 
> Hi @vkucera , which variables are you referring to? If I am not wrong, the variable names in the `HFXicToPKPiCandidateSelector.cxx` are equal to the ones in `HFLcToPKPiCandidateSelector.cxx`, but I am not sure if you are referring to them.

I meant the configurables and histograms, but it's fine, we can rename them later, together with everything else.

vkucera reviewed (dismissed):
@ginnocen Good from my side.

Commit:
PWGHF: pair Xic code with adaptAnalysisTask and charge() changes

mfaggin commented:
@vkucera I followed your instructions and the `adaptAnalysisTask` and `charge()` changes should be correctly in place.

vkucera commented on Analysis/Tasks/PWGHF/HFXicToPKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,447 @@
+  template <typename T>
+  bool daughterSelection(const T& track)
+  {
+    if (track.sign() == 0) {
+      return false;
+    }
```
If I see it correctly, the track charge sign is checked only here.
Since the track charge can never be 0, this condition is useless and can be removed.

mfaggin replied:
At the moment yes, there are no other places where the charge is checked.
For the moment, I followed your instructions.

A good thing would be to check the sign of the 3 daughter together, verifying if they are like +-+ or -+- (in other words: avoiding to build triplets with +++, ---, --+ and so on).
But I prefer to think at it for the next PR eventually.

vkucera replied:
Where do you get these triplets? From the skimming you get only +-+ or -+-.

mfaggin replied:
... ok, then what I was thinking is already done. Thanks for the clarification.

Commit:
PWGHF: remove check for charge to be different from 0

vkucera commented on Analysis/Tasks/PWGHF/taskXic.cxx:
```diff
@@ -0,0 +1,171 @@
+      }
+      if (std::abs(candidate.flagMCMatchRec()) == XicToPKPi) {
+        // Get the corresponding MC particle.
+        auto indexMother = RecoDecay::getMother(particlesMC, candidate.index0_as<aod::BigTracksMC>().label_as<soa::Join<aod::McParticles, aod::HfCandProng3MCGen>>(), 4232, true);
```
`label` -> `mcParticle`

mfaggin replied:
@vkucera the change is in place

Commit:
PWGHF: change label_as into mcParticle_as in taskXic

vkucera reviewed (dismissed):
@ginnocen

ginnocen commented on Analysis/Tasks/PWGHF/taskXic.cxx:
```diff
@@ -0,0 +1,171 @@
+     {"hd0Prong0", "3-prong candidates;prong 0 DCAxy to prim. vertex (cm);entries", {HistType::kTH1F, {{100, -1., 1.}}}},
+     {"hd0Prong1", "3-prong candidates;prong 1 DCAxy to prim. vertex (cm);entries", {HistType::kTH1F, {{100, -1., 1.}}}},
+     {"hd0Prong2", "3-prong candidates;prong 1 DCAxy to prim. vertex (cm);entries", {HistType::kTH1F, {{100, -1., 1.}}}},
+     {"hCt", "3-prong candidates;proper lifetime (#Lambda_{c}) * #it{c} (cm);entries", {HistType::kTH1F, {{120, -20., 100.}}}},
```
Shouldn't this be Xi_c @mfaggin?

mfaggin replied:
@ginnocen I fixed the particle name

Commit:
PWGHF: fix particle label for ct distribution in taskXic

ginnocen commented on Analysis/Tasks/PWGHF/taskXic.cxx:
```diff
@@ -0,0 +1,171 @@
+     {"hEtaGen", "MC particles (matched);#it{#eta};entries", {HistType::kTH1F, {{100, -2., 2.}}}}}};
+
+  Configurable<int> d_selectionFlagXic{"d_selectionFlagXic", 1, "Selection Flag for Xic"};
+  Configurable<int> d_selectionFlagXicbar{"d_selectionFlagXicbar", 1, "Selection Flag for Xicbar"};
```
This d_selectionFlagXicbar is obsolete i think, probably a leftover. I would remove it 
(same for Lc) @mfaggin @DelloStritto

mfaggin replied:
@ginnocen I removed that line

ginnocen commented:
hi @mfaggin @vkucera. I added a couple of very minor comments. We will need to improve the selector, as you know :). Once those two little details are solved I think it is ready to go. Thanks to both for the nice work.

Commit:
PWGHF: remove selection flag for Xicbar (obsolete) from taskXic

mfaggin committed:
Merge branch 'dev' into mfaggin_XicTopKpi

jgrosseo commented:
Mac CI pending since 2 days. Merging.
